File: fastapi-backend/services/cache.py
import time
from collections import OrderedDict

class LRUCache:
    def __init__(self, capacity: int, ttl_seconds: int):
        self.cache = OrderedDict()
        self.capacity = capacity
        self.ttl = ttl_seconds

    def get(self, key: str):
        if key not in self.cache:
            return None
        
        value, timestamp = self.cache[key]
        if time.time() - timestamp > self.ttl:
            del self.cache[key]
            return None
            
        # Move to end to show it was recently used
        self.cache.move_to_end(key)
        print("[CACHE HIT] - saved ~250ms")
        return value

    def set(self, key: str, value: str):
        if key in self.cache:
            self.cache.move_to_end(key)
        self.cache[key] = (value, time.time())
        if len(self.cache) > self.capacity:
            self.cache.popitem(last=False)

# Global singleton with 200 items max and 120s TTL
global_cache = LRUCache(200, 120)

File: backend/src/services/cache.js
const LRUCache = class {
  constructor(maxSize, ttlStr) {
    this.maxSize = maxSize;
    this.ttl = ttlStr;
    this.cache = new Map();
  }

  get(key) {
    if (!this.cache.has(key)) return null;
    
    const entry = this.cache.get(key);
    if (Date.now() - entry.timestamp > this.ttl) {
      this.cache.delete(key);
      return null;
    }
    
    // Move to most recently used
    this.cache.delete(key);
    this.cache.set(key, entry);
    
    console.log(`[CACHE HIT] - saved ~250ms`);
    return entry.value;
  }

  set(key, value) {
    if (this.cache.has(key)) {
      this.cache.delete(key);
    } else if (this.cache.size >= this.maxSize) {
      // Delete least recently used (first item)
      const firstKey = this.cache.keys().next().value;
      this.cache.delete(firstKey);
    }
    this.cache.set(key, { value, timestamp: Date.now() });
  }
};

// Global singleton cache instance with 200 entries and 120s TTL
const globalCache = new LRUCache(200, 120000);

module.exports = globalCache;
